evap/rewards/views.py

```diff
@@ -13,76 +13,58 @@
 from evap.evaluation.tools import AttachmentResponse, get_object_from_dict_pk_entry_or_logged_40x
 from evap.rewards.exporters import RewardsExporter
-from evap.rewards.forms import RewardPointRedemptionEventForm
+from evap.rewards.forms import RewardPointRedemptionEventForm, RewardPointRedemptionFormSet
 from evap.rewards.models import (
-    NoPointsSelectedError,
-    NotEnoughPointsError,
-    OutdatedRedemptionDataError,
-    RedemptionEventExpiredError,
     RewardPointGranting,
     RewardPointRedemption,
     RewardPointRedemptionEvent,
     SemesterActivation,
 )
-from evap.rewards.tools import grant_eligible_reward_points_for_semester, reward_points_of_user, save_redemptions
-
-
-def redeem_reward_points(request):
-    redemptions = {}
-    try:
-        for key, value in request.POST.items():
-            if key.startswith("points-"):
-                event_id = int(key.rpartition("-")[2])
-                redemptions[event_id] = int(value)
-        previous_redeemed_points = int(request.POST["previous_redeemed_points"])
-    except (ValueError, KeyError, TypeError) as e:
-        raise BadRequest from e
-
-    try:
-        save_redemptions(request, redemptions, previous_redeemed_points)
-        messages.success(request, _("You successfully redeemed your points."))
-    except (
-        NoPointsSelectedError,
-        NotEnoughPointsError,
-        RedemptionEventExpiredError,
-        OutdatedRedemptionDataError,
-    ) as error:
-        status_code = 400
-        match error:
-            case NoPointsSelectedError():
-                error_string = _("You cannot redeem 0 points.")
-            case NotEnoughPointsError():
-                error_string = _("You don't have enough reward points.")
-            case RedemptionEventExpiredError():
-                error_string = _("Sorry, the deadline for this event expired already.")
-            case OutdatedRedemptionDataError():
-                status_code = 409
-                error_string = _(
-                    "It appears that your browser sent multiple redemption requests. You can see all successful redemptions below."
-                )
-            case _:
-                assert_never(type(error))
-
-        messages.error(request, error_string)
-        return status_code
-    return 200
+from evap.rewards.tools import grant_eligible_reward_points_for_semester, redeemed_points_of_user, reward_points_of_user
 
 
 @reward_user_required
 def index(request):
     status = 200
-    if request.method == "POST":
-        status = redeem_reward_points(request)
+
+    events = RewardPointRedemptionEvent.objects.filter(redeem_end_date__gte=datetime.now().date()).order_by("date")
+
+    # pylint: disable=unexpected-keyword-arg
+    formset = RewardPointRedemptionFormSet(
+        request.POST or None,
+        initial=[{"event": e, "points": 0} for e in events],
+        user=request.user,
+    )
+    with formset.lock():
```

Comment:
This seems a bit fragile, I would much rather have something where `formset` is created inside a `with lock_for_reward_changes(request.user)` and never escapes it, but I see that this conflicts with the Django form workflow. Thoughts?

I now wonder whether `.lock()` is the best name here, because the formset is not being locked, but rather we touch some rows in the database to include them in the transaction that we have made atomic. From my current understanding, `formset.atomic()` would better fit this. What do you think?

Reply:
> where formset is created inside a with lock_for_reward_changes(request.user) and never escapes it, but I see that this conflicts with the Django form workflow.

We can have the context manager be useable as a class thing instead of a instance thing, have it notify its locking globally. I don't see why this wouldn't work. Don't know if its more elegant, though.

No opinion on the name. `lock`, `assert`, `critical_section`, whatever, all fine with me. I think "context within which we validate correctness of requested redemptions and apply them" is too long, so I don't see a name that wouldn't require people to look up what exactly it does when modifying the code.

Reply:
> We can have the context manager be useable as a class thing instead of a instance thing, have it notify its locking globally.

I think I would prefer that, it would probably look okay with the form escaping the with-block:
```python
with lock_for_reward_points_upgrade(request.user):
    formset = ...
    if ...:
        formset.save()
        return ...

return render(..., formset)
```

Reply:
@hansegucker pointed out that this would make the assertions in `clean` and `save` more awkward - I didn't know that these were a requirement, it's fair though.

I now thought that maybe a good way to do it would be something like
```python
class TheFormset:
    @classmethod
    @contextmanager
    @transaction.atomic
    def make_in_transaction(cls, *args, **kwargs):
        list(...) # lock the stuff
        instance = cls(*args, **kwargs)
        instance.locked = True
        try:
            yield instance
        finally:
            instance.locked = False
```

Reply:
> @hansegucker pointed out that this would make the assertions in clean and save more awkward - I didn't know that these were a requirement, it's fair though.

Why? I'd have thought we just set and unset a global/class variable, instead of an instance variable.

> I now thought that maybe a good way to do it would be something like

Also fine with your approach